### DISCUSSION TITLE
If my custom filter were to turn this codeblock into a figure, how would I do that? and keep the codeblock's attributes?

### DISCUSSION THREAD

**Discussion by danmackinlay** [quarto-dev/quarto-cli#8926] [Q&A] [answered]
### Description

I developed a [custom filter to render TikZ as SVG](https://github.com/quarto-dev/quarto-cli/discussions/8925), available [here](https://github.com/danmackinlay/quarto_tikz?tab=readme-ov-file) including [an MWE](https://github.com/danmackinlay/quarto_tikz/blob/main/example.qmd).

The filter successfully renders graphics, so that is nice. However, its integration with quarto is not entirely harmonious. Let us look at an example in the MWE:

````
```{.tikz .test-class #fig-example embed_mode="link" scale=3 filename="example" format="svg" caption="Rendering TikZ DAGs is easy"}
\node[draw, circle] (A) at (0,0) {A};
\node[draw, circle] (B) at (2,2) {$\phi$};
\node[draw, circle] (C) at (4,0) {C};
\node[draw, circle] (seven) at (5,1) {888888};
\draw[->] (A) -- (B);
\draw[->] (B) -- (C);
\draw[->] (C) -- (A);
```
````

Now, the HTML generated by the filter in the MWE is not quite what I would like:

```html
<div class="code-with-filename">
  <div class="quarto-figure quarto-figure-center">
    <figure class="figure">
      <p><img src="./images/example-1.svg" class="img-fluid figure-img"></p>
      <figcaption>Rendering TikZ DAGs is easy</figcaption>
    </figure>
  </div>
</div>
```

The native format output is also suspicious

```haskell
Figure
      ( "tikz" , [] , [] )
      (Caption
         Nothing [ Plain [ Str "Rendering TikZ DAGs is easy" ] ])
      [ Plain
          [ Image
              ( "" , [] , [] ) [] ( "./images/example-1.svg" , "" )
          ]
      ]
```
Several, presumably-related problems are visible here:

1. the codeblock div (represented by `div.code-with-filename` in HTML is still there, even though the filter tries very hard to replace it rather than append to it; that is kind of messy formatting-wise.
2. neither the `div` nor the `figure` have managed to retain their class or id attributes, even though we try to set it in the function [`tikz_walker`](https://github.com/danmackinlay/quarto_tikz/blob/main/_extensions/tikz/tikz.lua) by doing `local figure = pandoc.Figure({ image }, caption, cb.classes, cb.identifier)`
3. as such this image cannot be linked to. It is not a crossreferenceable Figure.

Can the wise creators of the lua API advise if what I am attempting is possible, and if so, what I have done wrong in the current version?

**Comment by mcanouil** [comment]
Were you aware of <https://github.com/pandoc-ext/diagram>?

**Comment by cscheid** [comment]
Cross-posting [my answer](https://github.com/quarto-dev/quarto-cli/discussions/8925#discussioncomment-8623848):

That kind of thing is why we came up with the div syntax for cross-references and the new, more permissive structure of Quarto 1.4 crossref's system. What happens if you try the following?

````

::: {#fig-tikz}

```{.tikz ...}
...
```

This is a caption
:::

See @fig-tikz.
````

**Reply by danmackinlay** [reply]
oops My attempt to keep the discussion categories separated may not have worked. Sorry to  lure you into cross posting @cscheid .
If I render this:

````qmd
::: {#fig-example .test-class}
```{.tikz  embed_mode="link" scale=3 filename="example" format="svg" caption="Rendering TikZ DAGs is easy"}
\node[draw, circle] (A) at (0,0) {A};
\node[draw, circle] (B) at (2,2) {$\phi$};
\node[draw, circle] (C) at (4,0) {C};
\node[draw, circle] (seven) at (5,1) {888888};
\draw[->] (A) -- (B);
\draw[->] (B) -- (C);
\draw[->] (C) -- (A);
```
:::
````
I get this

````html
<div id="fig-example" class="test-class quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="code-with-filename">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/example-1.svg" class="img-fluid figure-img"></p>
<figcaption>Rendering TikZ DAGs is easy</figcaption>
</figure>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1
</figcaption>
</figure>
<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="#fig-example" style="font: 1em / 1 anchorjs-icons; margin-left: 0.1875em; padding-right: 0.1875em; padding-left: 0.1875em;"></a></div>
````

So now it's a referenceable figure, and also the caption syntax is nicer, so that's good. There is still the `<div class="code-with-filename">` wrapping the inner content, which is messy but I guess probably cosmetic only.

I guess that leads me to two questions:
1. should the filter attempt to simply output a pandoc.Image and leave all the figure stuff to quarto div-handling?
2. what is ... happening with the enclosing block element in filter output? I'm finding it hard to understand the processing of the output of the filter in terms of pandoc's lua API. Why do identifiers and classes vanish from the outermost object? Is the behvaiour as-expected but under-documented, or a bug?

**Reply by cscheid** [reply]
> Is the behvaiour as-expected but under-documented, or a bug?

it's as expected; I don't know what you believe to be the right amount of documentation here, so I'll simply say it's just not doing what you expect. The enclosing block element is being converted into one of Quarto's custom AST elements (in this case, a FloatRefTarget), so you can write Lua filters that operate on them directly. If you want an enclosing div around your content, you should write that explicitly:

````
::: {#fig-tikz}

::: {.test-class}
```{.tikz ...}
...
```
:::

A caption
:::

See @fig-tikz.
````

We don't in general put a div around every emitted custom AST element; the enclosing block is simply declaring the node and enclosing the content. Each node can process the classes and attributes as it sees fit. In this case, the output is correct for a variety of reasons:

- that enclosing div encloses everything including the caption. It would be ambiguous whether to forward the classes to a div enclosing the content instead (what if the user intended it to surround the caption? what if they _didn't_? There's no correct answer, so we do neither and ask you to be explicit)
- some output formats have no concept of blocks, but do have the concept of a Figure. What would forwarding the class do in that case?

> There is still the `<div class="code-with-filename">` wrapping the inner content, which is messy but I guess probably cosmetic only.

Somehow we're interpreting your output to be a code block; I think your filter is running too late in our pipeline, after Quarto has detected your code block and processed it.

**Reply by danmackinlay** [reply]
thank you this is clarifying @cscheid I'm grateful for your thoughtful responses.
I think you've given me the info I was looking for but let me see if I can assemble it into the answer to the question I was trying to frame::

A block element X I return in my pandoc lua filter via might be subject to arbitrary replacement or processing by quarto after I return it. If I wish X itself to possess classes or an identifier, that is not an option. X must rather be enclosed  in a wrapping div with the desired identifier or classes, or X must itself enclose some other element with the desired identifier or class. Is that about right?

In summary:

````qmd
::: {#fig-with-id .these .classes .will .persist}

::: {.these .classes .are .ok #this-id-is-ok}
```{.tikz .all .my .classes .will .be .deleted .because .i .will .be .filtered #ids-too}
...
```
:::

[A caption]{.i .can .also .have .classes #and-ids}
:::

See @fig-with-id.
````

regarding the pipeline order, is it not the case that [per default user filters run before quarto filters?](https://quarto.org/docs/extensions/filters.html#activating-filters)?

**Reply by cscheid** [reply]
> is it not the case that [per default user filters run before quarto filters?](https://quarto.org/docs/extensions/filters.html#activating-filters)?

It is.

Oh, ok, now I know what's happening. It's the presence of the `filename` attribute in a code block. That triggers a _different_ custom AST for us (a "DecoratedCodeBlock", and _that_ one is not broadly documented.) So your filter is finding a CodeBlock element and handling it, but that happens to be the content of the DecoratedCodeBlock, which is why you see the `<div class="code-with-filename">` element.

This is a bit unfortunate. You can handle that entire element wholesale, by filtering against `DecoratedCodeBlock` as well as `CodeBlock`. The `DecoratedCodeBlock` node has a `code_block` field as well as a `filename` field. You should be able to do something like

```lua

CodeBlock = function(code)
--- your handler here
end

DecoratedCodeBlock = function(node)
  return CodeBlock(node.code_block)
end
```

**Reply by danmackinlay** [reply]
I've realised that part of my problem was that I was looking at the [Lua API docs](https://quarto.org/docs/extensions/lua-api.html) which does not include the [Quarto 1.4 updates to the Lua API](https://quarto.org/docs/prerelease/1.4/lua_changes.html#custom-formats-and-custom-renderers); I was able to find them thats the @cscheid 's metnion of the magical FloatRefTarget target, thanks!

That enables me to instantly answer my second question from the previous post: 

>regarding the pipeline order, is it not the case that [per default user filters run before quarto filters?](https://quarto.org/docs/extensions/filters.html#activating-filters)?

WRONG, that is merely what the lua API docs say. [checking the 1.4 update](https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases) we see that there are multiple phases now: `ast`, `quarto`, and `render`, so I am free to experiment with putting my filter in any of them.
.

**Reply by danmackinlay** [reply]
Oh cripes, thanks for that update @cscheid I didn't notice response arrive while I was typing mine. Sorry, I hope my "WRONG" doesn't sound too brusque, that was directed at me not you ;-)

**Reply by danmackinlay** [reply]
Your DecoratedCodeBlock trick seems to work, FWIW. Also, I could simply *not* use the "filename" key

**Reply by cscheid** [reply]
> WRONG, that is merely what the lua API docs say. [checking the 1.4 update](https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases) we see that there are multiple phases now: ast, quarto, and render, so I am free to experiment with putting my filter in any of them.

The lua API docs are incomplete, but correct; by default, your filter does happen exactly where it did in 1.3. We just give you more options now.

**Comment by danmackinlay** [comment]
That does leave me with a new question I cannot quite glean from the docs:

if I (or [data-intuitive/quarto-d2/](https://github.com/data-intuitive/quarto-d2/) or [pandoc-ext/diagram](https://github.com/pandoc-ext/diagram/tree/main)) wish to emit a figure, is the recommended method now,

1. Don't. Instead, emit an `Image`, and let the user mark it as a Figure via the fancy-div method, OR
2. Emit the custom quarto AST [`FloatRefTarget`](https://quarto.org/docs/prerelease/1.4/lua_changes.html#the-floatreftarget-ast-node), constructing that element and whatever internal structure it has and returning that?

**Reply by cscheid** [reply, marked as answer]
You can do either. If you do the latter in Lua, though, you need to use the constructor `quarto.FloatRefTarget { ... }`. It should work, but I don't think that's well documented right now (only so many hours in the day...)

**Reply by danmackinlay** [reply]
That was a real tour-de-force @cscheid  thank you.

I should note that this means that extensions that emit generic Figures are going to have a rough ride at the moment; this include @mcanouil 's suggested figure renderer [pandoc-ext/diagram](https://github.com/pandoc-ext/diagram/)

**Reply by danmackinlay** [reply]
I went to open a ticket about hat and discovered they know already: https://github.com/pandoc-ext/diagram/issues/13

**Reply by cscheid** [reply]
Aside from the nested figure business I mentioned, they should work and generate content that's cross-referenceable (but maybe that's you mean by rough ride).
